DOC: Add clearer info when copy is False but memory shared only for certain objects

mdhsieh opened the pull request:
- [ ] closes #41423
- [ ] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

Added to Series section more info in `copy` constructor parameter and 2 examples.

Ran validation script:
```
python scripts/validate_docstrings.py pandas.Series
```

Output:
```
################################################################################
########################## Docstring (pandas.Series)  ##########################
################################################################################

One-dimensional ndarray with axis labels (including time series).

Labels need not be unique but must be a hashable type. The object
supports both integer- and label-based indexing and provides a host of
methods for performing operations involving the index. Statistical
methods from ndarray have been overridden to automatically exclude
missing data (currently represented as NaN).

Operations between Series (+, -, /, *, **) align values based on their
associated index values-- they need not be the same length. The result
index will be the sorted union of the two indexes.

Parameters
----------
data : array-like, Iterable, dict, or scalar value
    Contains data stored in Series. If data is a dict, argument order is
    maintained.
index : array-like or Index (1d)
    Values must be hashable and have the same length as `data`.
    Non-unique index values are allowed. Will default to
    RangeIndex (0, 1, 2, ..., n) if not provided. If data is dict-like
    and index is None, then the keys in the data are used as the index. If the
    index is not None, the resulting Series is reindexed with the index values.
dtype : str, numpy.dtype, or ExtensionDtype, optional
    Data type for the output Series. If not specified, this will be
    inferred from `data`.
    See the :ref:`user guide <basics.dtypes>` for more usages.
name : str, optional
    The name to give to the Series.
copy : bool, default False
    Copy input data. If False and the Series returns a `copy`
    of the data, the memory location for the values is not shared.
    If False and the Series returns a `view` on the data,
    the memory location for the values is shared.

Examples
--------
Constructing Series from a dictionary with an Index specified

>>> d = {'a': 1, 'b': 2, 'c': 3}
>>> ser = pd.Series(data=d, index=['a', 'b', 'c'])
>>> ser
a   1
b   2
c   3
dtype: int64

The keys of the dictionary match with the Index values, hence the Index
values have no effect.

>>> d = {'a': 1, 'b': 2, 'c': 3}
>>> ser = pd.Series(data=d, index=['x', 'y', 'z'])
>>> ser
x   NaN
y   NaN
z   NaN
dtype: float64

Note that the Index is first build with the keys from the dictionary.
After this the Series is reindexed with the given Index values, hence we
get all NaN as a result.

Constructing Series from an array with `copy=False`.

>>> r = [1,2]
>>> ser = pd.Series(r, copy=False)
>>> ser.iloc[0] = 999
>>> r
[1, 2]
>>> ser
0    999
1      2
dtype: int64

The Series returns a `copy` of the original data, so
the original data is unchanged.

Constructing Series from a `numpy.array` with `copy=False`.

>>> r = np.array([1,2])
>>> ser = pd.Series(r, copy=False)
>>> ser.iloc[0] = 999
>>> r
array([999,   2])
>>> ser
0    999
1      2
dtype: int32

The Series returns a `view` on the original data, so
the original data is changed as well.

################################################################################
################################## Validation ##################################
################################################################################

3 Errors found:
        Parameters {'fastpath'} not documented
        See Also section not found
        flake8 error: E902 PermissionError: [Errno 13] Permission denied: 'C:\\Users\\mdhsi\\AppData\\Local\\Temp\\tmpxz50buet'
```

mdhsieh committed:
DOC: Add info on memory location sharing when copy=False (#41423).

Update is on main Series documentation.

mdhsieh committed:
DOC: Add 2 examples when creating Series and copy is False (#41423).

Original data unchanged when Series is a Copy,
but original data is changed when Series is a View.

mdhsieh committed:
DOC: Fix code format of 2 examples added to Series section (#41423).

After building documentation, found the code examples were not displayed correctly.
Added newlines and rebuilt to fix.

pep8speaks commented:
Hello @mdhsieh! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pandas/core/series.py`](https://github.com/pandas-dev/pandas/blob/2ed2b8dbdec90cb0f8060a9423e49f7078db3ed9/pandas/core/series.py):

> [Line 257:1](https://github.com/pandas-dev/pandas/blob/2ed2b8dbdec90cb0f8060a9423e49f7078db3ed9/pandas/core/series.py#L257): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 259:1](https://github.com/pandas-dev/pandas/blob/2ed2b8dbdec90cb0f8060a9423e49f7078db3ed9/pandas/core/series.py#L259): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 269:1](https://github.com/pandas-dev/pandas/blob/2ed2b8dbdec90cb0f8060a9423e49f7078db3ed9/pandas/core/series.py#L269): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 272:1](https://github.com/pandas-dev/pandas/blob/2ed2b8dbdec90cb0f8060a9423e49f7078db3ed9/pandas/core/series.py#L272): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 274:1](https://github.com/pandas-dev/pandas/blob/2ed2b8dbdec90cb0f8060a9423e49f7078db3ed9/pandas/core/series.py#L274): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 284:1](https://github.com/pandas-dev/pandas/blob/2ed2b8dbdec90cb0f8060a9423e49f7078db3ed9/pandas/core/series.py#L284): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

mdhsieh commented:
@github-actions pre-commit

mzeitlin11 commented on pandas/core/series.py:
```diff
@@ -251,6 +254,36 @@ class Series(base.IndexOpsMixin, generic.NDFrame):
     After this the Series is reindexed with the given Index values, hence we
     get all NaN as a result.
+    
+    Constructing Series from an array with `copy=False`.
```
I'd call this a list, since the difference between an array and list in this context is important in terms of ability to share memory.

attack68 replied:
```suggestion
    Constructing Series from a list with `copy=False`.
```

mzeitlin11 commented on pandas/core/series.py:
```diff
@@ -251,6 +254,36 @@ class Series(base.IndexOpsMixin, generic.NDFrame):
+    1      2
+    dtype: int64
+    
+    The Series returns a `copy` of the original data, so
```
What do you mean by returns here? It's more like the data backing the series does not share the same memory as the input `r`.

mdhsieh replied:
Oh right, by "returns" a copy I meant the Series created had a copy of the data. I guess "returns" is not the correct word to use.

mzeitlin11 commented on pandas/core/series.py:
```diff
@@ -251,6 +254,36 @@ class Series(base.IndexOpsMixin, generic.NDFrame):
+    
+    Constructing Series from an array with `copy=False`.
+    
+    >>> r = [1,2]
```
```suggestion
    >>> r = [1, 2]
```

mzeitlin11 commented on pandas/core/series.py:
```diff
@@ -251,6 +254,36 @@ class Series(base.IndexOpsMixin, generic.NDFrame):
+    
+    Constructing Series from a `numpy.array` with `copy=False`.
+    
+    >>> r = np.array([1,2])
```
```suggestion
    >>> r = np.array([1, 2])
```

mzeitlin11 commented on pandas/core/series.py:
```diff
@@ -223,7 +223,10 @@ class Series(base.IndexOpsMixin, generic.NDFrame):
         The name to give to the Series.
     copy : bool, default False
-        Copy input data.
+        Copy input data. If False and the Series returns a `copy`
```
Same comment on usage of "returns" as below. 

Generally I think this description might be too verbose for a parameter description and could be shortened to basically just get the idea across that a copy may still be made if necessary (as in the case of a list where we cannot share memory). Then the examples below should suffice to explain the different possible cases.

mzeitlin11 replied:
(https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html has a good example of this)

mzeitlin11 commented on pandas/core/series.py:
```diff
@@ -251,6 +254,36 @@ class Series(base.IndexOpsMixin, generic.NDFrame):
+    dtype: int32
+    
+    The Series returns a `view` on the original data, so
+    the original data is changed as well.
```
Same comment as above on wording of returns

mzeitlin11 reviewed (commented):
Thanks for helping to clarify this behavior @mdhsieh! Left some comments below

attack68 commented on pandas/core/series.py:
```diff
@@ -223,7 +223,10 @@ class Series(base.IndexOpsMixin, generic.NDFrame):
         The name to give to the Series.
     copy : bool, default False
-        Copy input data.
+        Copy input data. If False and the Series returns a `copy`
+        of the data, the memory location for the values is not shared.
+        If False and the Series returns a `view` on the data,
+        the memory location for the values is shared.
```
```suggestion
        Copy input data. Only affects Series or 1d ndarray input. See examples.
```

attack68 commented on pandas/core/series.py:
```diff
@@ -251,6 +254,36 @@ class Series(base.IndexOpsMixin, generic.NDFrame):
+    1      2
+    dtype: int64
+    
+    The Series returns a `copy` of the original data, so
```
```suggestion
    Due to input data type the Series is created with a `copy` even though `copy=False`, so
```

attack68 commented on pandas/core/series.py:
```diff
@@ -251,6 +254,36 @@ class Series(base.IndexOpsMixin, generic.NDFrame):
+    1      2
+    dtype: int32
+    
+    The Series returns a `view` on the original data, so
```
```suggestion
    Due to input data type here the Series represents a `view` on the original data, so
```

attack68 reviewed (changes requested):
good start, very similar comments to @mzeitlin11 and suggetsed changes.

mdhsieh committed:
DOC: Fix spacing and reword parameter info and examples (#41423).

Remove return, replace with has.
Remove array, replace with list.
Fix array spacing to match other Series examples.
Replace long param info with only affects sentence, based off DataFrame documentation.
Reword example explanation.
Change numpy.array to 1d ndarray to match other documentation, like DataFrame.

mdhsieh committed:
DOC: Try fix PEP8 spacing issue.

mdhsieh committed:
Merge remote-tracking branch 'upstream/master' into 41423-doc-series-copy

mdhsieh committed:
DOC: Try fix one line which is too long.

mdhsieh committed:
Revert "Merge remote-tracking branch 'upstream/master' into 41423-doc-series-copy"

This reverts commit 5eab8f9967d632d7d2500a3cf1b82d3cc3fbce4e, reversing
changes made to 671cf86a78e931a9c98ad72571ec65cd3c35d8a7.

Need to revert merge commit to find out why build is failing after pushing branches.

These were the commands run to do the revert, at previous commit:
check parent branches of the merge commit

git cat-file  -p 5eab8f9967d632d7d2500a3cf1b82d3cc3fbce4e

then revert commit

git revert --no-commit -m 2 5eab8f9967d632d7d2500a3cf1b82d3cc3fbce4e

Then accepted incoming change in series.py, which removed my info and examples added.

mdhsieh committed:
DOC: Add reworded parameter info and and 2 examples again (#41423).

Remove whitespace on blank lines and line break sentences to avoid style errors.
Previous commit had 1 failing check.

MarcoGorelli commented on pandas/core/series.py:
```diff
@@ -251,6 +251,38 @@ class Series(base.IndexOpsMixin, generic.NDFrame):
+
+    Constructing Series from a 1d ndarray with `copy=False`.
+
+    >>> r = np.array([1, 2])
+    >>> ser = pd.Series(r, copy=False)
+    >>> ser.iloc[0] = 999
+    >>> r
+    array([999,   2])
+    >>> ser
+    0    999
+    1      2
+    dtype: int32
```
looks like this test is failing in CI:
```
=================================== FAILURES ===================================
_____________________ [doctest] pandas.core.series.Series ______________________
269     the data is unchanged.
270 
271     Constructing Series from a 1d ndarray with `copy=False`.
272 
273     >>> r = np.array([1, 2])
274     >>> ser = pd.Series(r, copy=False)
275     >>> ser.iloc[0] = 999
276     >>> r
277     array([999,   2])
278     >>> ser
Differences (unified diff with -expected +actual):
    @@ -1,3 +1,3 @@
     0    999
     1      2
    -dtype: int32
    +dtype: int64
```

mdhsieh replied:
Oh got it, I didn't see that. Appreciate the help.

mdhsieh committed:
DOC: Change type output from int32 to int64 in ndarray data Series code example.

mdhsieh commented:
Thanks @mzeitlin11 @attack68 . I've tried rewording as suggested.

attack68 reviewed (approved):
looks clearer to me. thanks.

jreback commented:
thanks @mdhsieh
